apply_license.py

```diff
@@ -76,76 +75,79 @@ def match_patterns(s, patterns):
             return True
     return False
 
+def skip_path(fullpath):
+    return match_patterns(fullpath, excluded_paths)
 
-def force_include(fullpath):
-    if match_patterns(fullpath, include_patterns):
-        return True
-    return False
-
-def skip_filename(fullpath, basename):
-    if basename in exluded_files:
-        return True
-
-    if match_patterns(fullpath, exluded_patterns):
-        return True
+def skip_filename(fullpath):
+    return match_patterns(fullpath, excluded_files)
 
-    return False
+def has_defold_license(s):
+    return re.search(RE_LICENSE, s, flags=re.DOTALL) is not None
 
 def has_other_license(s):
-    return "Copyright" in s or "License" in s
+    return ("Copyright" in s or "License" in s) and not has_defold_license(s)
 
-def get_comment_style_for_file(fullpath):
-    ext = os.path.splitext(fullpath)[1]
-    if ext == ".go" and not fullpath.startswith("./go/"):
+def get_license_for_file(filepath):
+    ext = os.path.splitext(filepath)[1]
+    if not ext in ext_to_license:
         return None
-    return ext_to_comment.get(ext)
+    return ext_to_license[ext]
+
+def apply_license(license, contents):
+    # Preserve shebang
+    if contents.startswith("#!"):
+        firstline = contents.partition('\n')[0]
+        contents = contents.replace(firstline, "")
+        license = firstline + "\n" + license
+    return license  + "\n\n" + contents
 
+def check_ignored(path):
+    return subprocess.call(['git', 'check-ignore', '-q', path]) == 0
 
 
-for subdir, dirs, files in os.walk("."):
-    # Ignore tmp and build folders
-    if subdir.startswith("./tmp") or "build" in subdir:
-        continue
+
+dryrun = False
+
+for root, dirs, files in os.walk(".", topdown=True):
+    # exclude dirs to avoid traversing them at all
+    # with topdown set to True we can make in place modifications of dirs to
+    # have os.walk() skip directories
+    dirs[:] = [ d for d in dirs if not skip_path(root + os.sep + d) and not check_ignored(root + os.sep + d) ]
+
     for file in files:
-        filepath = subdir + os.sep + file
-        ext = os.path.splitext(filepath)[1]
-        if ext not in extensions:
-            continue
+        filepath = root + os.sep + file
```

Comment:
Fyi, there's also `os.path.join(a,b)`

Reply:
True. Not sure why I decided on os.sep instead of os.path.join(). Changed it!